flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -142,6 +187,13 @@ public void checkpointState(
 		// We generally try to emit the checkpoint barrier as soon as possible to not affect downstream
 		// checkpoint alignments
 
+		// Step (0): Record the last triggered checkpointId.
+		lastCheckpointId = metadata.getCheckpointId();
```

Comment:
`lastCheckpointId = Math.max(lastCheckpointId, metadata.getCheckpointId());`
?

Reply:
`checkpointState` is executed in sync phase of checkpoint on task which should behave as a strictly increasing order due to the barrier align mechanism.

Reply:
It can also be triggered through RPC (for sources).
I agree that it shouldn't be possible now, but I think it's good to have such a guard or check to protect against bugs or changes.

Reply:
If so, the condition below have to change which lead to I have to record a lot of unnecessary checkpoint ids.

~~~java
if (checkpointId > lastCheckpointId) {
	// only record checkpoints that have not triggered on task side.
       abortedCheckpointIds.add(checkpointId);
}
~~~

Reply:
We can just force the assumption with
```Preconditions.checkArgument(lastCheckpointId < metadata.getCheckpointId());```